### ISSUE TITLE
Getting multiple snapshots

### ISSUE THREAD

**Issue by mabushey** [jimsalterjrs/sanoid#328]
amber/music@autosnap_2019-01-01_00:00:00_yearly                                          0      -  3.93T  -
amber/music@autosnap_2019-01-01_00:00:00_monthly                                         0      -  3.93T  -
amber/music@autosnap_2019-01-01_00:01:00_yearly                                          0      -  3.93T  -
amber/music@autosnap_2019-01-01_00:01:00_monthly                                         0      -  3.93T  -

`crontab -l`
`* * * * * TZ=UTC /usr/local/bin/sanoid --cron >> /var/log/sanoid.log`

`/etc/sanoid/sanoid.conf`:
```
[amber/music]
        use_template = music 
...
[template_music]
        hourly = 5
        daily = 4
        monthly = 3
        yearly = 5
        autosnap = yes
        autoprune = yes
```

This was running an older version (1.8?). I upgraded to 2.0.0 earlier today.

**Comment by phreaker0** [comment]
I guess the the sanoid execution time is higher than a minute and the first run isn't finished when cron schedules the next run after a minute, so the second sanoid run uses the old cache file despite the first sanoid run already created the snapshots but didn't update the cache file yet.

To solve this, you need to make sure that only one instance of sanoid runs at a time. I use systemd timers for that purpose as they by default only run a single instance. For cron you could use a tool like "flock" to get the same behaviour.

**Comment by shodanshok** [comment]
Try converting your crontab into something similar:

`*/5 * * * * TZ=UTC /usr/local/bin/sanoid --cron >> /var/log/sanoid.log`

This would run `sanoid` each *five* minutes, which should provide it ample time to finish.

**Comment by mabushey** [comment]
Thanks for the replies. :) I set the cron to run every 5 minutes. I once had an issue (on Linux) where a custom go app was launching an application three times in parallel (instead of once). I added a bash launcher that used flock, and found out the flock is NOT atomic, flock launched all three. I switched to creating a lock directory (this is atomic) and this worked. Basically flock doesn't do what it's designed to do correctly. It would work in this application however.

**Comment by jimsalterjrs** [comment]
We actually do use a file lock system when we're creating snapshots, but it 
only locks during the snapshot creation itself. On a sufficiently busy 
system, it's possible to have it lock for the first instance, then unlock 
before the second instance asks to take a snapshot; if the second instance 
was launched prior to the first instance updating the cache, it'll be 
operating on stale data and mistakenly think it still needs to take a snapshot.

Fixing this has been a pretty low priority for me personally because it 
doesn't hurt anything; the snapshots get taken, and the extra unnecessary 
snapshots consume almost no space since they're taken so closely with the 
first ones. :)

**Comment by mabushey** [comment]
I went with flock after all, I've had issues in the past where a raidz2 fills completely up and becomes so unresponsive that even 5 minutes probably wouldn't cut it. I used:
```
* * * * * TZ=UTC flock -n /tmp/sandroid /usr/local/bin/sanoid --cron --verbose >> /var/log/sanoid.log
```
